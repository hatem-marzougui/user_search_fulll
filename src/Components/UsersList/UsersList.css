.users-list-container {
  width: 100%;
  height: calc(100vh - 64px - 10vh - 50px - 50px);
  border: 1px solid red;
  display: grid;
  grid-template-columns: repeat(4, 1fr);
  grid-gap: 15px;
  overflow-y: auto;
  padding: 30px 30px;
  box-sizing: border-box;
  background-color: #f0f0f0;
  position: relative;
}
.message {
  position: absolute;
  top: 50%;
  left: 50%;
  transform: translate(-50%, -50%);
  color: rgb(103, 92, 107);
  font-size: 1.2rem;
  font-weight: bold;
  text-align: center;
}
