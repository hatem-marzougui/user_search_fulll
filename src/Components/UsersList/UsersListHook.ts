import { useMemo, useState } from "react";
import { getUsers } from "../../api/usersApi";
import { debounce } from "../../utils/debounce";
import {
  EMPTY_SEARCH_QUERY_ERROR,
  RATE_LIMIT,
  RATE_LIMIT_ERROR_MESSAGE,
} from "../../Constants";
import { User } from "../../types/user";

export const useUsersList = () => {
  const [githubUsers, setGithubUsers] = useState<User[]>([]);
  const [selectedItems, setSelectedItems] = useState<string[]>([]);
  const [editMode, setEditMode] = useState<boolean>(false);
  const [searchQuery, setSearchQuery] = useState<string>("");
  const [loading, setLoading] = useState<boolean>(false);
  const [error, setError] = useState<string | null>(EMPTY_SEARCH_QUERY_ERROR);

  const fetchUsers = async (query: string) => {
    if (!query.trim()) {
      // If the search query is empty, reset the users list without making an API call
      setGithubUsers([]);
      setError(EMPTY_SEARCH_QUERY_ERROR);
      return;
    }
    setLoading(true);
    setError(null); // Reset error state before fetching

    try {
      const data = await getUsers(query);
      setGithubUsers(data.items || []);
    } catch (err) {
      const message = (err as Error).message;
      if (message.includes(RATE_LIMIT)) {
        setError(RATE_LIMIT_ERROR_MESSAGE);
        setGithubUsers([]); // Clear users on rate limit error
      } else setError((err as Error).message);
    } finally {
      setLoading(false);
    }
  };
  const debouncedFetchUsers = useMemo(() => debounce(fetchUsers, 300), []);
  const handleSearch = (e: React.ChangeEvent<HTMLInputElement>) => {
    setSearchQuery(e.target.value);
    debouncedFetchUsers(e.target.value); // Debounce the search function
    // fetchUsers(e.target.value);
    setSelectedItems([]); // Reset selected items on new search
  };
  const handleToggleSelect = (e: React.ChangeEvent<HTMLInputElement>) => {
    setSelectedItems(
      selectedItems.includes(e.target.value)
        ? selectedItems.filter((item) => item !== e.target.value)
        : [...selectedItems, e.target.value]
    );
  };

  const handleSelectAllUsers = (e: React.ChangeEvent<HTMLInputElement>) => {
    if (e.target.checked) {
      setSelectedItems(githubUsers.map((user) => user.id.toString()));
    } else {
      setSelectedItems([]); // Clear selection if unchecked
    }
  };
  const handleDeleteUsers = () => {
    setGithubUsers(
      githubUsers.filter((user) => !selectedItems.includes(user.id.toString()))
    );
    setSelectedItems([]); // Reset selected items after deletion
  };
  const handleDuplicateUsers = () => {
    const duplicatedUsers = githubUsers
      .filter((user) => selectedItems.includes(user.id.toString()))
      .map((user) => ({
        ...user,
        id: Math.random(),
      }));
    setGithubUsers([...duplicatedUsers, ...githubUsers]);
    //   setSelectedItems([]); // Reset selected items after duplication
  };

  const handleEditMode = () => {
    setEditMode(!editMode);
    setSelectedItems([]); // Reset selected items when toggling edit mode
  };

  return {
    githubUsers,
    searchQuery,
    editMode,
    selectedItems,
    error,
    loading,
    handlers: {
      handleSearch,
      handleToggleSelect,
      handleDeleteUsers,
      handleDuplicateUsers,
      handleSelectAllUsers,
      handleEditMode,
    },
  };
};
