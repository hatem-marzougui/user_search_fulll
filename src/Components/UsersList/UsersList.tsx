import "./UsersList.css";
import { useUsersList } from "./UsersListHook";
import UserCard from "./UserCard/UserCard";
import Search from "./Search/Search";
import { NO_RESULTS_FOUND } from "../../Constants";
import ListControls from "./ListControls/ListControls";

const UsersList = () => {
  const {
    githubUsers,
    editMode,
    searchQuery,
    selectedItems,
    error,
    loading,
    handlers,
  } = useUsersList();

  return (
    <>
      <Search onSearchChange={handlers.handleSearch} value={searchQuery} />

      <ListControls
        selectedItems={selectedItems}
        handleDuplicateUsers={handlers.handleDuplicateUsers}
        handleDeleteUsers={handlers.handleDeleteUsers}
        githubUsers={githubUsers}
        editMode={editMode}
        handleEditMode={handlers.handleEditMode}
      />
      <div className="users-list-container" data-testid="users-list">
        {!loading && (error || githubUsers.length === 0) && (
          <div className="message">{error ? `${error}` : NO_RESULTS_FOUND}</div>
        )}

        {githubUsers.length > 0 && editMode && (
          <input
            className="select-all-checkbox"
            type="checkbox"
            onChange={handlers.handleSelectAllUsers}
            checked={
              selectedItems.length === githubUsers.length &&
              githubUsers.length > 0
            }
          />
        )}
        {githubUsers.map((user, index) => (
          <UserCard
            key={index}
            user={user}
            handleToggleSelect={handlers.handleToggleSelect}
            isSelected={selectedItems.includes(user.id.toString())}
            isEditMode={editMode}
          />
        ))}
      </div>
    </>
  );
};

export default UsersList;
