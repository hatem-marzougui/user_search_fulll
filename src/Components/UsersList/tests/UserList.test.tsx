import { render, screen } from "@testing-library/react";
import UsersList from "../UsersList";

jest.mock("../UsersListHook", () => ({
  useUsersList: () => ({
    githubUsers: [
      {
        id: "1",
        login: "hatem",
        avatar_url: "https://example.com/avatar.jpg",
      },
    ],
    editMode: false,
    searchQuery: "",
    selectedItems: [],
    error: null,
    loading: false,
    handlers: {
      handleSearch: jest.fn(),
      handleDuplicateUsers: jest.fn(),
      handleDeleteUsers: jest.fn(),
      handleEditMode: jest.fn(),
      handleToggleSelect: jest.fn(),
      handleSelectAllUsers: jest.fn(),
    },
  }),
}));

describe("UsersList", () => {
  test("renders UserCard when githubUsers is not empty", () => {
    render(<UsersList />);
    expect(screen.getByText(/hatem/i)).toBeInTheDocument();
  });

  test("matches snapshot", () => {
    const { container } = render(<UsersList />);
    expect(container).toMatchSnapshot();
  });
});
