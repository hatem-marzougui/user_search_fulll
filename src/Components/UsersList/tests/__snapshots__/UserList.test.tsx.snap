// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UsersList matches snapshot 1`] = `
<div>
  <div
    class="search-container"
  >
    <input
      class="search-input"
      placeholder="Rechercher..."
      type="search"
      value=""
    />
  </div>
  <div
    class="control-div"
  >
    <div
      class="users-count"
    >
      <img
        alt="liste"
        class="icon"
        height="24"
        src="/icons/list.svg"
        width="24"
      />
      <span>
        0
         
        elements selected
      </span>
    </div>
    <div
      class="controls-buttons"
    >
      <label
        class="switch"
      >
        <input
          type="checkbox"
        />
        <span
          class="slider round"
        />
      </label>
      <span>
        Edit Mode
         
      </span>
    </div>
  </div>
  <div
    class="users-list-container"
    data-testid="users-list"
  >
    <div
      class="user-card"
    >
      <img
        alt="Avatar"
        class="avatar"
        src="https://example.com/avatar.jpg"
      />
      <h6>
        ID: 
        1
      </h6>
      <h6>
        Login
        : 
        hatem
      </h6>
      <button
        class="viewProfile_button"
      >
        View Profile
      </button>
    </div>
  </div>
</div>
`;
